File: src/core/Shared/Adapters/Options/Options.cs
#nullable enable
using System;
using System.Linq;
using core.Options;

namespace core.Adapters.Options
{
    public class OptionChain
    {
        public OptionChain(string symbol, decimal volatility, int numberOfContracts, OptionDetail[] options)
        {
            Symbol = symbol;
            Volatility = volatility;
            NumberOfContracts = numberOfContracts;
            Options = options;
        }
        public string Symbol { get; }
        public decimal Volatility { get; }
        public int NumberOfContracts { get; }
        public OptionDetail[] Options { get;}

        public OptionDetail? FindMatchingOption(decimal strikePrice, string expirationDate, OptionType optionType)
        {
            return Options.FirstOrDefault(
                o => o.StrikePrice == strikePrice
                && o.ExpirationDate == expirationDate
                && o.OptionType == optionType.ToString().ToLowerInvariant());
        }
    }

    public class OptionDetail
    {
        public OptionDetail(string symbol, string side, string description)
        {
            Symbol = symbol;
            Side = side;
            Description = description;
        }

        public string Symbol { get; }
        public string Description { get; }
        public string ExpirationDate => ParsedExpirationDate.ToString("yyyy-MM-dd");
        public decimal StrikePrice { get; set; }
        public string Side { get; set; }
        public long Volume { get; set; }
        public long OpenInterest { get; set; }
        public decimal Bid { get; set; }
        public decimal Ask { get; set; }
        public string OptionType => Side;
        public bool IsCall => Side == "call";
        public bool IsPut => Side == "put";
        public decimal Spread => Ask - Bid;
        public decimal Volatility { get; set; }
        public decimal Delta { get; set; }
        public decimal Gamma { get; set; }
        public decimal Theta { get; set; }
        public decimal Vega { get; set; }
        public decimal Rho { get; set; }
        public decimal TimeValue { get; set; }
        public decimal IntrinsicValue { get; set; }
        public bool InTheMoney { get; set; }
        public string? ExchangeName { get; set; }
        public int DaysToExpiration { get; set; }
        public decimal MarkChange { get; set; }
        public decimal MarkPercentChange { get; set; }
        public decimal? UnderlyingPrice { get; set; }

        public decimal PerDayPrice
        {
            get
            {
                var date = ParsedExpirationDate;

                var today = DateTime.UtcNow.Date;

                var diff = Math.Max((int)date.Subtract(today).TotalDays, 1);

                return Bid * 100 / diff;
            }
        }

        public decimal BreakEven
        {
            get
            {
                return IsCall ?
                    StrikePrice + Bid
                    : StrikePrice - Bid;
            }
        }

        public decimal Risk => Bid / StrikePrice;

        public DateTimeOffset ParsedExpirationDate { get; set; }
    }
}

#nullable restore

File: src/core.fs/Adapters/Options.fs
namespace core.fs.Adapters.Options

open System
open core.fs.Options

type OptionDetail(symbol:string, optionType:OptionType, description:string, expiration:OptionExpiration) =
    
    member this.Symbol = symbol
    member this.Description = description
    member this.Expiration: OptionExpiration = expiration
    member val StrikePrice: decimal = 0m with get, set
    member val Volume: int64 = 0L with get, set
    member val OpenInterest: int64 = 0L with get, set
    member val Bid: decimal = 0m with get, set
    member val Ask: decimal = 0m with get, set
    member val Last: decimal = 0m with get, set
    member val Mark: decimal = 0m with get, set
    member this.OptionType = optionType
    member this.IsCall = match optionType with | Call -> true | _ -> false
    member this.IsPut = match optionType with | Put -> true | _ -> false
    member this.Spread = this.Ask - this.Bid
    member val Volatility: decimal = 0m with get, set
    member val Delta: decimal = 0m with get, set
    member val Gamma: decimal = 0m with get, set
    member val Theta: decimal = 0m with get, set
    member val Vega: decimal = 0m with get, set
    member val Rho: decimal = 0m with get, set
    member val TimeValue: decimal = 0m with get, set
    member val IntrinsicValue: decimal = 0m with get, set
    member val InTheMoney: bool = false with get, set
    member val ExchangeName: string option = None with get, set
    member val DaysToExpiration: int = 0 with get, set
    member val MarkChange: decimal = 0m with get, set
    member val MarkPercentChange: decimal = 0m with get, set
    member val UnderlyingPrice: decimal option = None with get, set
    
    member this.PerDayPrice =
        
        let date = this.Expiration.ToDateTimeOffset().Date
        let today = DateTimeOffset.UtcNow.Date
        let diff = Math.Max(int(date.Subtract(today).TotalDays), 1)
        this.Bid * 100m / decimal(diff)
        
    member this.BreakEven =
        
        if this.IsCall then
            this.StrikePrice + this.Bid
        else
            this.StrikePrice - this.Bid
            
    member this.Risk =
        
        this.Bid / this.StrikePrice


type OptionChain(symbol: string, volatility: decimal, numberOfContracts: decimal, options: OptionDetail[], underlyingPrice: decimal option) =
    member this.Symbol = symbol
    member this.Volatility = volatility
    member this.NumberOfContracts = numberOfContracts
    member this.Options = options
    member this.UnderlyingPrice = underlyingPrice

    member this.FindMatchingOption(strikePrice: decimal, expirationDate: OptionExpiration, optionType: OptionType) =
        options
        |> Seq.tryFind(fun o -> o.StrikePrice = strikePrice && o.Expiration = expirationDate && o.OptionType = optionType)
